index.ts

```diff
@@ -0,0 +1,486 @@
+
+type Dict<T> = Record<string, T>;
+type PathPiece = string | number | Array<string | number>;
+
+function lastItem<T>(arr: Array<T>): T {
+  return arr[arr.length-1]
+}
+
+interface Parsed {
+  auth: string;
+  username: string;
+  password: string;
+  host: string;
+  hostname: string;
+  hash: string | null;
+  search: string;
+  pathname: string;
+  port: number | null;  // some protocols have no port, like file://
+  path: string;
+  query: string;
+  protocol: string;
+}
+
+function urlParse(href: string) {
+
+  if (!href.includes("://")) {
+    if (href.charAt(0) === "/") {
+      href = "http://localhost/"
+    }
+    href = "http://" + href
+  }
+  
+  const url = new URL(href)
+  const colonFreeProtocol = url.protocol.substring(0, url.protocol.length -1)
+  const hash = (url.hash && url.hash.charAt(0)) === "#" ? url.hash.substring(1) : url.hash
+
+  const result: Parsed = {
+    protocol: colonFreeProtocol,
+    auth: `${url.username}:${url.password}`,
+    username: url.username,
+    password: url.password,
+    host: url.host,
+    hostname: url.hostname,
+    hash,
+    search: url.search,
+    pathname: url.pathname,
+    port: url.port ? parseInt(url.port, 10) : null,
+    path: (url.pathname || "") + (url.search || ""),
+    query: url.search ? url.search.slice(1) : url.search,
+  }
+  return result
+}
+
+const getDefaults = function () {
+  const defaultUrl = "http://localhost:80"
+  const defaults = urlParse(defaultUrl)
+  return defaults
+}
+
+export class UrlGrey {
+  #url: string
+  _parsed: Parsed | null
+
+  constructor(url: string) {
+    this.#url = url
+    this._parsed = null
+  }
+
+  parsed() {
+    if (this._parsed) {
+      return this._parsed
+    }
+    this._parsed = urlParse(this.#url)
+    const defaults = getDefaults()
+    
+    this._parsed.protocol = this._parsed.protocol || defaults.protocol
+    return this._parsed
```

Comment:
We should mark this as `readonly` in typescript, or copy it on the way out. readonly is probably better. if we don't do one of these then calling code can change the internal state by mutating the return value.